## Logistic Regression
Logistic regression aims to solve classification problems. It does this by predicting categorical outcomes, unlike linear regression that predicts a continuous outcome.

In the simplest case there are two outcomes, which is called binomial, an example of which is predicting if a tumor is malignant or benign. Other cases have more than two outcomes to classify, in this case it is called multinomial. A common example for multinomial logistic regression would be predicting the class of an iris flower between 3 different species.

Here we will be using basic logistic regression to predict a binomial variable. This means it has only two possible outcomes.

In [ ]:
import np as np
from sklearn import linear_model

- We have predicted that a tumor with a size of 3.46mm will not be cancerous.

In [5]:
X = np.array([3.78, 2.44, 2.09, 0.14, 1.72, 1.65, 4.92, 4.37, 4.96, 4.52, 3.69, 5.88]).reshape(-1,1)
y = np.array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1])

logr = linear_model.LogisticRegression()
logr.fit(X,y)

predicted = logr.predict(np.array([3.46]).reshape(-1,1))

print(predicted)

[0]


## Coefficient
In logistic regression the coefficient is the expected change in log-odds of having the outcome per unit change in X.

This does not have the most intuitive understanding so let's use it to create something that makes more sense, odds.

- This tells us that as the size of a tumor increases by 1mm the odds of it being a cancerous tumor increases by 4x.

In [ ]:
X = np.array([3.78, 2.44, 2.09, 0.14, 1.72, 1.65, 4.92, 4.37, 4.96, 4.52, 3.69, 5.88]).reshape(-1,1)
y = np.array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1])

logr = linear_model.LogisticRegression()
logr.fit(X,y)

log_odds = logr.coef_ 
odds = np.exp(log_odds)

print(odds)

[[4.03557295]]


## Probability
The coefficient and intercept values can be used to find the probability that each tumor is cancerous.

Create a function that uses the model's coefficient and intercept values to return a new value. This new value represents probability that the given observation is a tumor:

In [16]:
X = np.array([3.78, 2.44, 2.09, 0.14, 1.72, 1.65, 4.92, 4.37, 4.96, 4.52, 3.69, 5.88]).reshape(-1,1)
y = np.array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1])

logr = linear_model.LogisticRegression()
logr.fit(X,y)

def logit2prob(logr, X):
  log_odds = logr.coef_ * X + logr.intercept_
  odds = np.exp(log_odds)
  probability = odds / (1 + odds)
  return(probability)

threshold = 0.5
probs = logit2prob(logr, X)

for size, prob in zip(X.flatten(), probs.flatten()):
    print(f"Tumor size: {size:.2f} cm -> Probability of cancer: {prob:.2f}")
    if prob > threshold:
        print("→ Likely cancerous!\n")
    else:
        print("→ Likely benign.\n")

# print(logit2prob(logr, X))

# logr.predict_proba(X)


Tumor size: 3.78 cm -> Probability of cancer: 0.61
→ Likely cancerous!

Tumor size: 2.44 cm -> Probability of cancer: 0.19
→ Likely benign.

Tumor size: 2.09 cm -> Probability of cancer: 0.13
→ Likely benign.

Tumor size: 0.14 cm -> Probability of cancer: 0.01
→ Likely benign.

Tumor size: 1.72 cm -> Probability of cancer: 0.08
→ Likely benign.

Tumor size: 1.65 cm -> Probability of cancer: 0.07
→ Likely benign.

Tumor size: 4.92 cm -> Probability of cancer: 0.88
→ Likely cancerous!

Tumor size: 4.37 cm -> Probability of cancer: 0.78
→ Likely cancerous!

Tumor size: 4.96 cm -> Probability of cancer: 0.89
→ Likely cancerous!

Tumor size: 4.52 cm -> Probability of cancer: 0.81
→ Likely cancerous!

Tumor size: 3.69 cm -> Probability of cancer: 0.58
→ Likely cancerous!

Tumor size: 5.88 cm -> Probability of cancer: 0.97
→ Likely cancerous!



## Results Explained
3.78 0.61 The probability that a tumor with the size 3.78cm is cancerous is 61%.

2.44 0.19 The probability that a tumor with the size 2.44cm is cancerous is 19%.

2.09 0.13 The probability that a tumor with the size 2.09cm is cancerous is 13%.